src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java

```diff
@@ -176,6 +178,15 @@ public static Builder builder() {
     return new Builder();
   }
 
+  /**
+   * Add this platform to the given fingerprint.
+   */
+  public void addTo(Fingerprint f) {
+    f.addString(label.toString());
+    f.addNullableString(remoteExecutionProperties);
+    f.addInt(constraints.size());
```

Comment:
Fixed.

Reply:
I'm somewhat out of the loop here -- is there a guarantee that `getExecutionInfo().constraints()` (and `.label()`) cannot influence the result of the action? At least naively, a subclass could just call that method and do different things based on that.

Reply:
There's no "getExecutionInfo()" in this class.

The idea here is that if the constraints are changed, or if the remote execution properties are changed, the action probably needs to be re-run even if everything else (especially the generated command-line) is the same.

The constraints are specified in a rule as a list of constraint_value targets, then converted into a map in PlatformInfo (based on each constraint value's unique constraint setting) for ease of lookup.

There is the case where the constraint value is updated (possibly the constraint setting is changed), but not the platform, in which case re-analysis will update the Action, but here the key would be unchanged.

So, a solution: we just keep recursing downwards, as usual.

Reply:
Erm, I misspoke *and* I think I don't understand something really trivial about this. Sorry for being dense. Let me try again.

Any AbstractAction subclass can call `this.getExecutionPlatform().constraints()` and make decisions based on the resulting map and the `ConstraintValueInfo` values in it. These are currently not added to the action key, thus making it possible for the action to change behavior without changing the key. Ways out:
 1. Enforce that any change in `PlatformInfo.constraints` causes `PlatformInfo.remoteExecutionProperties` to also change
 2. If `PlatformInfo.constraints()` cannot influence the action, remove the reference to it
 3. If it can, add `.constraints()` to the action key (this is the trivially correct course of action)

Do I understand correctly that your last paragraph means that the constraint value can change without the remote execution info changing? This would imply that (2) is the solution, I guess?

Reply:
When the execution platform changes (whether the constraints are changing, or the remote platform properties), the action _might_ change or it might not. There's no good way to tell (and in the case of the remote execution properties, only the remote execution system knows), so we need to always have a new action key.

This presents the problem that spurious changes to the execution platform will drown the cache, but in practice I don't expect execution platforms to change often. the motivating problem that I am trying to fix (and which I realize doesn't have an issue opened yet) is that a user updated the remote execution properties, and then Bazel used the cached result instead of telling the remote execution system to re-run the action, which would presumably have given different results.

The best answer here is 3) (and is what I am trying to implement). Ideally, 4) would also be true, but as you say that's not actually enforceable.

Opened #4756 to track this.

Reply:
In that case, I shall approve this anticipating a followup change.